src/ripple/protocol/NegativeUNL.h

```diff
@@ -0,0 +1,92 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2021 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_PROTOCOL_NEGATIVE_UNL_H_INCLUDED
+#define RIPPLE_PROTOCOL_NEGATIVE_UNL_H_INCLUDED
+
+#include <ripple/protocol/LedgerEntryWrapper.h>
+//#include <ripple/protocol/STAccount.h>
+//#include <ripple/protocol/STAmount.h>
+
+namespace ripple {
+
+template <bool Writable>
+class NegUNLImpl final : public LedgerEntryWrapper<Writable>
+{
+private:
+    using Base = LedgerEntryWrapper<Writable>;
+    using SleT = typename Base::SleT;
+    using Base::wrapped_;
+
+
+    // Friend declarations of factory functions.
+    //
+    // For classes that contain factories we must declare the entire class
+    // as a friend unless the class declaration is visible at this point.
+    friend class ReadView;
+    friend class ApplyView;
+
+public:
+    // This constructor is used in Change::applyUNLModify() function.
+    // Keshava: Is it mandatory to use make_XYZ factory functions for this usecase?
+    // What are the perils of having a public constructor?
+    NegUNLImpl(std::shared_ptr<SleT>&& w) : Base(std::move(w))
+    {
+    }
+
+    // Conversion operator from NegUNLImpl<true> to NegUNLImpl<false>.
+    operator NegUNLImpl<true>() const
+    {
+        return NegUNLImpl<false>(
+            std::const_pointer_cast<std::shared_ptr<STLedgerEntry const>>(
+                wrapped_));
+    }
+
+    [[nodiscard]] bool
+        isFieldPresent(SField const& field) const {
```

Comment:
My understanding is that the `NegativeUNL` and `Amendments` ledger objects are singletons. Hence, there is possibility of re-writing some of these functions as `static` functions.

But I'm not sure if that is a good design decision.